wcfsetup/install/files/lib/system/form/builder/field/UploadFormField.class.php

```diff
@@ -360,29 +369,36 @@ public function updatedObject(array $data, IStorableObject $object, $loadValues
                 $method = "variable '" . \get_class($object) . "::$" . $this->getObjectProperty() . "'";
             }
 
-            if (\is_array($value)) {
-                $value = \array_map(function ($v) use ($method) {
-                    if (!\is_string($v) || !\file_exists($v)) {
-                        throw new \InvalidArgumentException(
-                            "The {$method} must return an array of strings with the file locations for field '{$this->getId()}'."
-                        );
+            // in any case the given value has to be a valid array
+            if (!\is_array($value)) {
+                throw new \InvalidArgumentException(
+                    "The {$method} must return an array of " . ($expectUploadFile ? "objects of class '" . UploadFile::class . "'" : "strings") . " with the valid file locations for field '{$this->getId()}'."
+                );
+            }
+
+            $value = \array_map(function ($v) use ($method, $expectUploadFile) {
```

Comment:
Overall this is very hard to read and understand. I dislike, that both value types are handled here. Maybe we can get two methods, which are handling the different inputs? That would be much easier to understand and read. And thus methods could be called within the if-else structure, which determines the source of the values.